README.md
This is a fork from https://github.com/vsergeev/u-msgpack-python.

Mostly, several optimizations were done in this fork to make the code run faster on pure Python
(but a bit less flexible -- see below).

Additionally, a basic client/server structure is provided in `umsgpack_conn.py`

See: https://github.com/vsergeev/u-msgpack-python for documentation on how to use umsgpack.

To check how to use the client/server, check the docstrigs and `__main__` of `umsgpack_conn.py`.

Note that `umsgpack.py` can be used standalone and `umsgpack_conn.py` just requires umsgpack.py.

Changes from the original version:

- The compatibility mode from u-msgpack-python has been removed.

- Subclasses from raw types aren't accepted (the types must be a basic type
  such int, long, float, str, and not a subclass).

## License

u-msgpack-python is MIT licensed. See the included `LICENSE` file for more details.



CHANGELOG.md
* Version 1.6.1
	* Several performance changes were done so it should run faster on pure Python.
	* A client/server structure is now provided on umsgpack_conn.

* Version 1.6
    * Fix wide character unicode string serialization. Thanks to cforger for the bug report and fix (https://github.com/vsergeev/u-msgpack-python/issues/8).
    * Add module docstrings.
    * Add module `version` tuple.
    * Add Python standard library style `dumps` and `loads` serialization/deserialization aliases.
    * Rename unpack exceptions `KeyNotPrimitiveException` -> `UnhashableKeyException` and `KeyDuplicateException` -> `DuplicateKeyException`. Add aliases for backwards compatibility.

* Version 1.5
    * Hide internal helper functions from module's exported names.
    * Make unit tests more portable among interpreters (CPython, PyPy).
    * Update documentation for PyPy support.

* Version 1.4
    * Add type checking to Ext class initialization.
    * Add support for Python 2.6.

* Version 1.2
    * Add compatibility mode to support old specification "raw" bytes msgpack type.

* Version 1.0
    * Initial release.



test_performance.py
# coding: utf-8
import sys
import timeit


def create_vectors(mod):
    single_test_vectors = [
        # None
        [ "nil", None, b"\xc0" ],
        # Booleans
        [ "bool false", False, b"\xc2" ],
        [ "bool true", True, b"\xc3" ],
        # + 7-bit uint
        [ "7-bit uint", 0x00, b"\x00" ],
        [ "7-bit uint", 0x10, b"\x10" ],
        [ "7-bit uint", 0x7f, b"\x7f" ],
        # - 5-bit int
        [ "5-bit sint", -1, b"\xff" ],
        [ "5-bit sint", -16, b"\xf0" ],
        [ "5-bit sint", -32, b"\xe0" ],
        # 8-bit uint
        [ "8-bit uint", 0x80, b"\xcc\x80" ],
        [ "8-bit uint", 0xf0, b"\xcc\xf0" ],
        [ "8-bit uint", 0xff, b"\xcc\xff" ],
        # 16-bit uint
        [ "16-bit uint", 0x100, b"\xcd\x01\x00" ],
        [ "16-bit uint", 0x2000, b"\xcd\x20\x00" ],
        [ "16-bit uint", 0xffff, b"\xcd\xff\xff" ],
        # 32-bit uint
        [ "32-bit uint", 0x10000, b"\xce\x00\x01\x00\x00" ],
        [ "32-bit uint", 0x200000, b"\xce\x00\x20\x00\x00" ],
        [ "32-bit uint", 0xffffffff, b"\xce\xff\xff\xff\xff" ],
        # 64-bit uint
        [ "64-bit uint", 0x100000000, b"\xcf" + b"\x00\x00\x00\x01" + b"\x00\x00\x00\x00" ],
        [ "64-bit uint", 0x200000000000, b"\xcf" + b"\x00\x00\x20\x00" + b"\x00\x00\x00\x00" ],
        [ "64-bit uint", 0xffffffffffffffff, b"\xcf" + b"\xff\xff\xff\xff" + b"\xff\xff\xff\xff" ],
        # 8-bit int
        [ "8-bit int", -33, b"\xd0\xdf" ],
        [ "8-bit int", -100, b"\xd0\x9c" ],
        [ "8-bit int", -128, b"\xd0\x80" ],
        # 16-bit int
        [ "16-bit int", -129, b"\xd1\xff\x7f" ],
        [ "16-bit int", -2000, b"\xd1\xf8\x30" ],
        [ "16-bit int", -32768, b"\xd1\x80\x00" ],
        # 32-bit int
        [ "32-bit int", -32769, b"\xd2\xff\xff\x7f\xff" ],
        [ "32-bit int", -1000000000, b"\xd2\xc4\x65\x36\x00" ],
        [ "32-bit int", -2147483648, b"\xd2\x80\x00\x00\x00" ],
        # 64-bit int
        [ "64-bit int", -2147483649, b"\xd3" + b"\xff\xff\xff\xff" + b"\x7f\xff\xff\xff" ],
        [ "64-bit int", -1000000000000000002, b"\xd3" + b"\xf2\x1f\x49\x4c" + b"\x58\x9b\xff\xfe" ],
        [ "64-bit int", -9223372036854775808, b"\xd3" + b"\x80\x00\x00\x00" + b"\x00\x00\x00\x00" ],
        # 64-bit float
        [ "64-bit float", 0.0, b"\xcb" + b"\x00\x00\x00\x00" + b"\x00\x00\x00\x00" ],
        [ "64-bit float", 2.5, b"\xcb" + b"\x40\x04\x00\x00" + b"\x00\x00\x00\x00" ],
        [ "64-bit float", float(10 ** 35), b"\xcb" + b"\x47\x33\x42\x61" + b"\x72\xc7\x4d\x82" ],
        # Fixstr String
        [ "fix string", u"", b"\xa0" ],
        [ "fix string", u"a", b"\xa1\x61" ],
        [ "fix string", u"abc", b"\xa3\x61\x62\x63" ],
        [ "fix string", u"a" * 31, b"\xbf" + b"\x61"*31 ],
        # 8-bit String
        [ "8-bit string", u"b" * 32, b"\xd9\x20" + b"b" * 32 ],
        [ "8-bit string", u"c" * 100, b"\xd9\x64" + b"c" * 100 ],
        [ "8-bit string", u"d" * 255, b"\xd9\xff" + b"d" * 255 ],
        # 16-bit String
        [ "16-bit string", u"b" * 256, b"\xda\x01\x00" + b"b" * 256 ],
        [ "16-bit string", u"c" * 65535, b"\xda\xff\xff" + b"c" * 65535 ],
        # 32-bit String
        [ "32-bit string", u"b" * 65536, b"\xdb\x00\x01\x00\x00" + b"b" * 65536 ],
        # Wide character String
        [ "wide char string", u"Allagbé", b"\xa8Allagb\xc3\xa9" ],
        [ "wide char string", u"По оживлённым берегам", b"\xd9\x28\xd0\x9f\xd0\xbe\x20\xd0\xbe\xd0\xb6\xd0\xb8\xd0\xb2\xd0\xbb\xd1\x91\xd0\xbd\xd0\xbd\xd1\x8b\xd0\xbc\x20\xd0\xb1\xd0\xb5\xd1\x80\xd0\xb5\xd0\xb3\xd0\xb0\xd0\xbc" ],
        # 8-bit Binary
        [ "8-bit binary", b"\x80" * 1, b"\xc4\x01" + b"\x80" * 1 ],
        [ "8-bit binary", b"\x80" * 32, b"\xc4\x20" + b"\x80" * 32 ],
        [ "8-bit binary", b"\x80" * 255, b"\xc4\xff" + b"\x80" * 255 ],
        # 16-bit Binary
        [ "16-bit binary", b"\x80" * 256, b"\xc5\x01\x00" + b"\x80" * 256 ],
        # 32-bit Binary
        [ "32-bit binary", b"\x80" * 65536, b"\xc6\x00\x01\x00\x00" + b"\x80" * 65536 ],
        # Fixext 1
        [ "fixext 1", mod.Ext(0x05, b"\x80"*1), b"\xd4\x05" + b"\x80"*1 ],
        # Fixext 2
        [ "fixext 2", mod.Ext(0x05, b"\x80"*2), b"\xd5\x05" + b"\x80"*2 ],
        # Fixext 4
        [ "fixext 4", mod.Ext(0x05, b"\x80"*4), b"\xd6\x05" + b"\x80"*4 ],
        # Fixext 8
        [ "fixext 8", mod.Ext(0x05, b"\x80"*8), b"\xd7\x05" + b"\x80"*8 ],
        # Fixext 16
        [ "fixext 16", mod.Ext(0x05, b"\x80"*16), b"\xd8\x05" + b"\x80"*16 ],
        # 8-bit Ext
        [ "8-bit ext", mod.Ext(0x05, b"\x80"*255), b"\xc7\xff\x05" + b"\x80"*255 ],
        # 16-bit Ext
        [ "16-bit ext", mod.Ext(0x05, b"\x80"*256), b"\xc8\x01\x00\x05" + b"\x80"*256 ],
        # 32-bit Ext
        [ "32-bit ext", mod.Ext(0x05, b"\x80"*65536), b"\xc9\x00\x01\x00\x00\x05" + b"\x80"*65536 ],
        # Empty Array
        [ "empty array", [], b"\x90" ],
        # Empty Map
        [ "empty map", {}, b"\x80" ],
    ]
    return single_test_vectors


def test_pack_single():

    class Checker(object):

        def __init__(self, mod):
            self.mod = mod
            self.vectors = create_vectors(mod)
            self.times = {}
            self.big_array = list(range(1000))
            self.big_array_pack = self.mod.packb(self.big_array)
            self.big_str_array = ['a'] * 1000

        def check_pack(self):
            for vec in self.vectors:
                self.mod.packb(vec[1])

        def check_unpack(self):
            for vec in self.vectors:
                self.mod.unpackb(vec[2])

        def check_big_int_arr(self):
            self.mod.packb(self.big_array)

        def check_big_str_arr(self):
            self.mod.packb(self.big_str_array)

        def check_unpack_big_arr(self):
            self.mod.unpackb(self.big_array_pack)

        def check_all(self):
            self.times['check_pack'] = timeit.timeit(self.check_pack, number=2000)
            self.times['check_unpack'] = timeit.timeit(self.check_unpack, number=2000)
            self.times['check_big_int_arr'] = timeit.timeit(self.check_big_int_arr, number=2000)
            self.times['check_big_str_arr'] = timeit.timeit(self.check_big_str_arr, number=1000)
            self.times['check_unpack_big_arr'] = timeit.timeit(self.check_unpack_big_arr, number=1000)

    try:
        import _original_umsgpack
    except:
        sys.stderr.write('_original_umsgpack must be available with the old version of the code.\n')
        return
    checker_original = Checker(_original_umsgpack)
    checker_original.check_all()

    import umsgpack
    checker_new = Checker(umsgpack)
    checker_new.check_all()


    checker_original.check_all()
    checker_new.check_all()

    sys.stderr.write('Times (note: smaller ratio is better)\n')
    for key, new in sorted(checker_new.times.items()):
        old = checker_original.times[key]
        sys.stderr.write('%s: old time: %.4f new time: %.4f ratio: %.4f\n' % (key, old, new, new / old))




if __name__ == '__main__':
    test_pack_single()


test_umsgpack_conn.py
'''
@license: MIT
@author: Fabio Zadrozny
'''
import unittest

from umsgpack_conn import ConnectionHandler, assert_waited_condition
import umsgpack_conn


class Test(unittest.TestCase):


    def test_conn(self):

        server_received = []
        server_handlers = []

        class ServerHandler(ConnectionHandler):

            def __init__(self, connection):
                ConnectionHandler.__init__(self, connection)
                server_handlers.append(self)

            def _handle_decoded(self, decoded):
                server_received.append(decoded)

            def send(self, obj):
                self.connection.sendall(self.pack_obj(obj))



        client_received = []

        class ClientHandler(ConnectionHandler):

            def __init__(self, connection):
                ConnectionHandler.__init__(self, connection)

            def _handle_decoded(self, decoded):
                client_received.append(decoded)



        server = umsgpack_conn.Server(ServerHandler)
        server.serve_forever('127.0.0.1', 0, block=False)
        port = server.get_port()

        client = umsgpack_conn.Client('127.0.0.1', port, ClientHandler)

        assert_waited_condition(lambda: len(server_handlers) == 1)

        client.send('test send')
        assert_waited_condition(lambda: len(server_received) > 0)
        self.assertEqual(['test send'], server_received)
        del server_received[:]

        server_handlers[0].send({'from sender': 'with love'})
        assert_waited_condition(lambda: len(client_received) > 0)
        self.assertEqual([{'from sender': 'with love'}], client_received)
        del client_received[:]

        send = [1234, 'ab'] * 8192
        server_handlers[0].send(send)
        assert_waited_condition(lambda: len(client_received) > 0)
        self.assertEqual([send], client_received)

        assert server.is_alive()
        server.shutdown()
        assert_waited_condition(lambda: not server.is_alive())


if __name__ == "__main__":
    # import sys;sys.argv = ['', 'Test.test_conn']
    unittest.main()


umsgpack.py
# u-msgpack-python v1.6 - vsergeev at gmail
# https://github.com/vsergeev/u-msgpack-python
#
# u-msgpack-python is a lightweight MessagePack serializer and deserializer
# module, compatible with both Python 2 and 3, as well CPython and PyPy
# implementations of Python. u-msgpack-python is fully compliant with the
# latest MessagePack specification.com/msgpack/msgpack/blob/master/spec.md). In
# particular, it supports the new binary, UTF-8 string, and application ext
# types.
#
# MIT License
#
# Copyright (c) 2013-2014 Ivan A. Sergeev
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
"""
u-msgpack-python v1.6 - vsergeev at gmail
https://github.com/vsergeev/u-msgpack-python

u-msgpack-python is a lightweight MessagePack serializer and deserializer
module, compatible with both Python 2 and 3, as well CPython and PyPy
implementations of Python. u-msgpack-python is fully compliant with the
latest MessagePack specification.com/msgpack/msgpack/blob/master/spec.md). In
particular, it supports the new binary, UTF-8 string, and application ext
types.

License: MIT
"""

# Changes on 1.6.1 by Fabio Zadrozny: Did a number of performance enhancements.
version = (1, 6, 1)
"Module version tuple"

import collections
import struct
import sys


_struct_unpack = struct.unpack
_struct_pack = struct.pack

_IS_PY3 = sys.version_info[0] == 3

try:
    _xrange
except NameError:
    _xrange = range


################################################################################

# Extension type for application-specific types and data
class Ext:
    """
    The Ext class facilitates creating a serializable extension object to store
    an application-specific type and data byte array.
    """

    def __init__(self, type, data):
        """
        Construct a new Ext object.

        Args:
            type: application-specific type integer from 0 to 127
            data: application-specific data byte array

        Raises:
            TypeError:
                Specified ext type is outside of 0 to 127 range.

        Example:
        >>> foo = umsgpack.Ext(0x05, b"\x01\x02\x03")
        >>> umsgpack.packb({u"special stuff": foo, u"awesome": True})
        '\x82\xa7awesome\xc3\xadspecial stuff\xc7\x03\x05\x01\x02\x03'
        >>> bar = umsgpack.unpackb(_)
        >>> print(bar["special stuff"])
        Ext Object (Type: 0x05, Data: 01 02 03)
        >>>
        """
        # Application ext type should be 0 <= type <= 127
        if not type.__class__ == int or not (type >= 0 and type <= 127):
            raise TypeError("ext type out of range")

        # Check data is type bytes
        elif _IS_PY3 and not data.__class__ == bytes:
            raise TypeError("ext data is not type \'bytes\'")

        elif not _IS_PY3 and not data.__class__ == str:
            raise TypeError("ext data is not type \'str\'")

        self.type = type
        self.data = data

    def __eq__(self, other):
        """
        Compare this Ext object with another for equality.
        """
        return (isinstance(other, self.__class__) and
                self.type == other.type and
                self.data == other.data)

    def __ne__(self, other):
        """
        Compare this Ext object with another for inequality.
        """
        return not self.__eq__(other)

    def __str__(self):
        """
        String representation of this Ext object.
        """
        s = "Ext Object (Type: 0x%02x, Data: " % self.type
        for i in _xrange(min(len(self.data), 8)):
            if i > 0:
                s += " "
            if isinstance(self.data[i], int):
                s += "%02x" % (self.data[i])
            else:
                s += "%02x" % ord(self.data[i])
        if len(self.data) > 8:
            s += " ..."
        s += ")"
        return s

################################################################################

# Base Exception classes
class PackException(Exception):
    "Base class for exceptions encountered during packing."
    pass
class UnpackException(Exception):
    "Base class for exceptions encountered during unpacking."
    pass

# Packing error
class UnsupportedTypeException(PackException):
    "Object type not supported for packing."
    pass

# Unpacking error
class InsufficientDataException(UnpackException):
    "Insufficient data to unpack the encoded object."
    pass
class InvalidStringException(UnpackException):
    "Invalid UTF-8 string encountered during unpacking."
    pass
class ReservedCodeException(UnpackException):
    "Reserved code encountered during unpacking."
    pass
class UnhashableKeyException(UnpackException):
    """
    Unhashable key encountered during map unpacking.
    The serialized map cannot be deserialized into a Python dictionary.
    """
    pass
class DuplicateKeyException(UnpackException):
    "Duplicate key encountered during map unpacking."
    pass

# Backwards compatibility
KeyNotPrimitiveException = UnhashableKeyException
KeyDuplicateException = DuplicateKeyException

################################################################################

# You may notice _struct_pack("B", x) instead of the simpler chr(x) in the code
# below. This is to allow for seamless Python 2 and 3 compatibility, as chr(x)
# has a str return type instead of bytes in Python 3, and _struct_pack(...) has
# the right return type in both versions.

def _pack_integer(x, write):
    if x < 0:
        if x >= -32:
            write(_struct_pack("b", x))
        elif x >= -2 ** (8 - 1):
            write(b"\xd0")
            write(_struct_pack("b", x))
        elif x >= -2 ** (16 - 1):
            write(b"\xd1")
            write(_struct_pack(">h", x))
        elif x >= -2 ** (32 - 1):
            write(b"\xd2")
            write(_struct_pack(">i", x))
        elif x >= -2 ** (64 - 1):
            write(b"\xd3")
            write(_struct_pack(">q", x))
        else:
            raise UnsupportedTypeException("huge signed int")
    else:
        if x <= 127:
            write(_struct_pack("B", x))
        elif x <= 2 ** 8 - 1:
            write(b"\xcc")
            write(_struct_pack("B", x))
        elif x <= 2 ** 16 - 1:
            write(b"\xcd")
            write(_struct_pack(">H", x))
        elif x <= 2 ** 32 - 1:
            write(b"\xce")
            write(_struct_pack(">I", x))
        elif x <= 2 ** 64 - 1:
            write(b"\xcf")
            write(_struct_pack(">Q", x))
        else:
            raise UnsupportedTypeException("huge unsigned int")

def _pack_nil(x, write):
    write(b"\xc0")

def _pack_boolean(x, write):
    write(b"\xc3" if x else b"\xc2")

# Auto-detect system float precision
if sys.float_info.mant_dig == 53:
    _float_size = 64

    def _pack_float(x, write):
        write(b"\xcb")
        write(_struct_pack(">d", x))
else:
    _float_size = 32

    def _pack_float(x, write):
        write(b"\xca")
        write(_struct_pack(">f", x))

def _pack_string(x, write):
    x = x.encode('utf-8')
    sz = len(x)

    if sz <= 31:
        write(_struct_pack("B", 0xa0 | sz))
        write(x)
    elif sz <= 2 ** 8 - 1:
        write(b"\xd9")
        write(_struct_pack("B", sz))
        write(x)
    elif sz <= 2 ** 16 - 1:
        write(b"\xda")
        write(_struct_pack(">H", sz))
        write(x)
    elif sz <= 2 ** 32 - 1:
        write(b"\xdb")
        write(_struct_pack(">I", sz))
        write(x)
    else:
        raise UnsupportedTypeException("huge string")

def _pack_binary(x, write):
    sz = len(x)
    if sz <= 2 ** 8 - 1:
        write(b"\xc4")
        write(_struct_pack("B", sz))
        write(x)
    elif sz <= 2 ** 16 - 1:
        write(b"\xc5")
        write(_struct_pack(">H", sz))
        write(x)
    elif sz <= 2 ** 32 - 1:
        write(b"\xc6")
        write(_struct_pack(">I", sz))
        write(x)
    else:
        raise UnsupportedTypeException("huge binary string")


def _pack_ext(x, write):
    sz = len(x.data)

    if sz == 1:
        write(b"\xd4")
        write(_struct_pack("B", x.type & 0xff))
        write(x.data)
    elif sz == 2:
        write(b"\xd5")
        write(_struct_pack("B", x.type & 0xff))
        write(x.data)
    elif sz == 4:
        write(b"\xd6")
        write(_struct_pack("B", x.type & 0xff))
        write(x.data)
    elif sz == 8:
        write(b"\xd7")
        write(_struct_pack("B", x.type & 0xff))
        write(x.data)
    elif sz == 16:
        write(b"\xd8")
        write(_struct_pack("B", x.type & 0xff))
        write(x.data)
    elif sz <= 2 ** 8 - 1:
        write(b"\xc7")
        write(_struct_pack("BB", sz, x.type & 0xff))
        write(x.data)
    elif sz <= 2 ** 16 - 1:
        write(b"\xc8")
        write(_struct_pack(">HB", sz, x.type & 0xff))
        write(x.data)
    elif sz <= 2 ** 32 - 1:
        write(b"\xc9")
        write(_struct_pack(">IB", sz, x.type & 0xff))
        write(x.data)
    else:
        raise UnsupportedTypeException("huge ext data")

def _pack_array(arr, write):
    sz = len(arr)

    if sz <= 15:
        write(_struct_pack("B", 0x90 | sz))
    elif sz <= 2 ** 16 - 1:
        write(b"\xdc")
        write(_struct_pack(">H", sz))
    elif sz <= 2 ** 32 - 1:
        write(b"\xdd")
        write(_struct_pack(">I", sz))
    else:
        raise UnsupportedTypeException("huge array")

    get = _pack_dispatch.__getitem__
    for x in arr:
        get(x.__class__)(x, write)

def _pack_array_(x, write):
    sz = len(x)

    if sz <= 15:
        write(_struct_pack("B", 0x90 | sz))
    elif sz <= 2 ** 16 - 1:
        write(b"\xdc")
        write(_struct_pack(">H", sz))
    elif sz <= 2 ** 32 - 1:
        write(b"\xdd")
        write(_struct_pack(">I", sz))
    else:
        raise UnsupportedTypeException("huge array")

    get = _pack_dispatch.__getitem__
    for e in x:
        get(e.__class__)(e, write)


if _IS_PY3:

    def _pack_map(x, write):
        sz = len(x)

        if sz <= 15:
            write(_struct_pack("B", 0x80 | sz))
        elif sz <= 2 ** 16 - 1:
            write(b"\xde")
            write(_struct_pack(">H", sz))
        elif sz <= 2 ** 32 - 1:
            write(b"\xdf")
            write(_struct_pack(">I", sz))
        else:
            raise UnsupportedTypeException("huge array")

        get = _pack_dispatch.__getitem__
        for k, v in x.items():
            get(k.__class__)(k, write)
            get(v.__class__)(v, write)
else:
    def _pack_map(x, write):
        sz = len(x)

        if sz <= 15:
            write(_struct_pack("B", 0x80 | sz))
        elif sz <= 2 ** 16 - 1:
            write(b"\xde")
            write(_struct_pack(">H", sz))
        elif sz <= 2 ** 32 - 1:
            write(b"\xdf")
            write(_struct_pack(">I", sz))
        else:
            raise UnsupportedTypeException("huge array")

        get = _pack_dispatch.__getitem__
        for k, v in x.iteritems():  # The only difference from the PY3 version is iteritems().
            get(k.__class__)(k, write)
            get(v.__class__)(v, write)


def packb(x):
    """
    Serialize a Python object into MessagePack bytes.

    Args:
        x: Python object

    Returns:
        A 'str' or 'bytes' containing the serialized bytes (depending on python version).

    Raises:
        UnsupportedType(PackException):
            Object type not supported for packing.

    Example:
    >>> umsgpack._packb({u"compact": True, u"schema": 0})
    '\x82\xa7compact\xc3\xa6schema\x00'
    >>>
    """

    # This is slower on py2.7 and 3.3 (on windows)
    # b = _BytesIO()
    # _packb(x, b.write)
    # return b.getvalue()

    lst = []
    try:
        _pack_dispatch[x.__class__](x, lst.append)
    except KeyError:
        raise UnsupportedTypeException("unsupported type: %s, %r" % (str(type(x)), x))
    return b''.join(lst)



################################################################################

def _unpack_integer(code, read_fn):
    if (ord(code) & 0xe0) == 0xe0:
        return _struct_unpack("b", code)[0]
    elif code == b'\xd0':
        return _struct_unpack("b", read_fn(1))[0]
    elif code == b'\xd1':
        return _struct_unpack(">h", read_fn(2))[0]
    elif code == b'\xd2':
        return _struct_unpack(">i", read_fn(4))[0]
    elif code == b'\xd3':
        return _struct_unpack(">q", read_fn(8))[0]
    elif (ord(code) & 0x80) == 0x00:
        return _struct_unpack("B", code)[0]
    elif code == b'\xcc':
        return _struct_unpack("B", read_fn(1))[0]
    elif code == b'\xcd':
        return _struct_unpack(">H", read_fn(2))[0]
    elif code == b'\xce':
        return _struct_unpack(">I", read_fn(4))[0]
    elif code == b'\xcf':
        return _struct_unpack(">Q", read_fn(8))[0]
    raise Exception("logic error, not int: 0x%02x" % ord(code))

def _unpack_reserved(code, read_fn):
    if code == b'\xc1':
        raise ReservedCodeException("encountered reserved code: 0x%02x" % ord(code))
    raise Exception("logic error, not reserved code: 0x%02x" % ord(code))

def _unpack_nil(code, read_fn):
    if code == b'\xc0':
        return None
    raise Exception("logic error, not nil: 0x%02x" % ord(code))

def _unpack_boolean(code, read_fn):
    if code == b'\xc2':
        return False
    elif code == b'\xc3':
        return True
    raise Exception("logic error, not boolean: 0x%02x" % ord(code))

def _unpack_float(code, read_fn):
    if code == b'\xca':
        return _struct_unpack(">f", read_fn(4))[0]
    elif code == b'\xcb':
        return _struct_unpack(">d", read_fn(8))[0]
    raise Exception("logic error, not float: 0x%02x" % ord(code))

def _unpack_string(code, read_fn):
    if (ord(code) & 0xe0) == 0xa0:
        length = ord(code) & ~0xe0
    elif code == b'\xd9':
        length = _struct_unpack("B", read_fn(1))[0]
    elif code == b'\xda':
        length = _struct_unpack(">H", read_fn(2))[0]
    elif code == b'\xdb':
        length = _struct_unpack(">I", read_fn(4))[0]
    else:
        raise Exception("logic error, not string: 0x%02x" % ord(code))

    try:
        return bytes.decode(read_fn(length), 'utf-8')
    except UnicodeDecodeError:
        raise InvalidStringException("unpacked string is not utf-8")

def _unpack_binary(code, read_fn):
    if code == b'\xc4':
        length = _struct_unpack("B", read_fn(1))[0]
    elif code == b'\xc5':
        length = _struct_unpack(">H", read_fn(2))[0]
    elif code == b'\xc6':
        length = _struct_unpack(">I", read_fn(4))[0]
    else:
        raise Exception("logic error, not binary: 0x%02x" % ord(code))

    return read_fn(length)

def _unpack_ext(code, read_fn):
    if code == b'\xd4':
        length = 1
    elif code == b'\xd5':
        length = 2
    elif code == b'\xd6':
        length = 4
    elif code == b'\xd7':
        length = 8
    elif code == b'\xd8':
        length = 16
    elif code == b'\xc7':
        length = _struct_unpack("B", read_fn(1))[0]
    elif code == b'\xc8':
        length = _struct_unpack(">H", read_fn(2))[0]
    elif code == b'\xc9':
        length = _struct_unpack(">I", read_fn(4))[0]
    else:
        raise Exception("logic error, not ext: 0x%02x" % ord(code))

    return Ext(ord(read_fn(1)), read_fn(length))

def _unpack_array(code, read_fn):
    if (ord(code) & 0xf0) == 0x90:
        length = (ord(code) & ~0xf0)
    elif code == b'\xdc':
        length = _struct_unpack(">H", read_fn(2))[0]
    elif code == b'\xdd':
        length = _struct_unpack(">I", read_fn(4))[0]
    else:
        raise Exception("logic error, not array: 0x%02x" % ord(code))


    # The code below was: return [_unpackb(read_fn) for _i in _xrange(length)]
    # On Py 2.7 the code below was around 7% faster (mostly inlining some things).
    ret = []
    get = _unpack_dispatch_table.__getitem__
    append = ret.append
    while length > 0:
        length -= 1
        code = read_fn(1)
        append(get(code)(code, read_fn))
    return ret

def _unpack_map(code, read_fn):
    if (ord(code) & 0xf0) == 0x80:
        length = (ord(code) & ~0xf0)
    elif code == b'\xde':
        length = _struct_unpack(">H", read_fn(2))[0]
    elif code == b'\xdf':
        length = _struct_unpack(">I", read_fn(4))[0]
    else:
        raise Exception("logic error, not map: 0x%02x" % ord(code))

    get = _unpack_dispatch_table.__getitem__

    d = {}
    setitem = d.__setitem__
    while length > 0:
        length -= 1
        key_code = read_fn(1)
        k = get(key_code)(key_code, read_fn)
        
        val_code = read_fn(1)
        setitem(k, get(val_code)(val_code, read_fn))
    return d

########################################
def _byte_reader(s):
    i = [0]
    len_s = len(s)
    def read_fn(n):
        j = i[0]
        if j + n > len_s:
            raise InsufficientDataException()
        substring = s[j:j + n]
        i[0] = j + n
        return substring
    return read_fn

def _unpackb(read_fn):
    code = read_fn(1)
    return _unpack_dispatch_table[code](code, read_fn)


if _IS_PY3:
    _byte_type = bytes
else:
    _byte_type = str

def unpackb(s):
    """
    Deserialize MessagePack bytes into a Python object.

    Args:
        s: a 'str' containing the MessagePack serialized bytes in Py 2 or 'bytes' in Py3.

    Returns:
        A deserialized Python object.

    Raises:
        TypeError:
            Packed data is not type 'str'.
        InsufficientDataException(UnpackException):
            Insufficient data to unpack the encoded object.
        InvalidStringException(UnpackException):
            Invalid UTF-8 string encountered during unpacking.
        ReservedCodeException(UnpackException):
            Reserved code encountered during unpacking.
        UnhashableKeyException(UnpackException):
            Unhashable key encountered during map unpacking.
            The serialized map cannot be deserialized into a Python dictionary.
        DuplicateKeyException(UnpackException):
            Duplicate key encountered during map unpacking.

    Example:
    >>> umsgpack.unpackb(b'\x82\xa7compact\xc3\xa6schema\x00')
    {u'compact': True, u'schema': 0}
    >>>
    """
    if s.__class__ != _byte_type:
        raise TypeError("packed data is not type '%s'" % (_byte_type,))
    read_fn = _byte_reader(s)
    return _unpackb(read_fn)


loads = unpackb
dumps = packb

# Dispatch table built in __init for fast lookup of unpacking function.
_unpack_dispatch_table = {}

# Dispatch table built in __init for fast lookup of packing function.
_pack_dispatch = {}

################################################################################

def __init():

    # Fix uint
    for code in _xrange(0, 0x7f + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_integer
    # Fix map
    for code in _xrange(0x80, 0x8f + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_map
    # Fix array
    for code in _xrange(0x90, 0x9f + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_array
    # Fix str
    for code in _xrange(0xa0, 0xbf + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_string
    # Nil
    _unpack_dispatch_table[b'\xc0'] = _unpack_nil
    # Reserved
    _unpack_dispatch_table[b'\xc1'] = _unpack_reserved
    # Boolean
    _unpack_dispatch_table[b'\xc2'] = _unpack_boolean
    _unpack_dispatch_table[b'\xc3'] = _unpack_boolean
    # Bin
    for code in _xrange(0xc4, 0xc6 + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_binary
    # Ext
    for code in _xrange(0xc7, 0xc9 + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_ext
    # Float
    _unpack_dispatch_table[b'\xca'] = _unpack_float
    _unpack_dispatch_table[b'\xcb'] = _unpack_float
    # Uint
    for code in _xrange(0xcc, 0xcf + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_integer
    # Int
    for code in _xrange(0xd0, 0xd3 + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_integer
    # Fixext
    for code in _xrange(0xd4, 0xd8 + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_ext
    # String
    for code in _xrange(0xd9, 0xdb + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_string
    # Array
    _unpack_dispatch_table[b'\xdc'] = _unpack_array
    _unpack_dispatch_table[b'\xdd'] = _unpack_array
    # Map
    _unpack_dispatch_table[b'\xde'] = _unpack_map
    _unpack_dispatch_table[b'\xdf'] = _unpack_map
    # Negative fixint
    for code in _xrange(0xe0, 0xff + 1):
        _unpack_dispatch_table[_struct_pack("B", code)] = _unpack_integer

    if not _IS_PY3:
        _pack_dispatch.update({
            bool: _pack_boolean,
            int:_pack_integer,
            long:_pack_integer,
            float: _pack_float,
            unicode: _pack_string,
            str:_pack_binary,
            list:_pack_array,
            tuple:_pack_array,
            dict:_pack_map,
            Ext:_pack_ext,
            None.__class__:_pack_nil
        })

    else:
        _pack_dispatch.update({
            bool: _pack_boolean,
            int:_pack_integer,
            float: _pack_float,
            str:_pack_string,
            bytes:_pack_binary,
            list:_pack_array,
            tuple:_pack_array,
            dict:_pack_map,
            Ext:_pack_ext,
            None.__class__:_pack_nil
        })


__init()
